### PR TITLE
kv/dlock: a package that implement distributed lock

### PR DESCRIPTION
**Pull request by tiancaiamao**
This package will be used by bootstrap later, to enable that only one TiDB do the work.

PTAL @c4pt0r  @coocood @shenli

### CODE REVIEW

File: kv/dlock/dlock.go [hunk 1]
```diff
@@ -0,0 +1,194 @@
...
+	}
+}
+
+// Unlock unlocks the resource.
+func (l *DLock) Unlock() {
+	l.unlockSignal <- struct{}{}
+}
+
+func (l *DLock) getLockKey() []byte {
+	lockKey := fmt.Sprintf("%s-LOCK", l.name)
```
Comment (coocood) [inline kv/dlock/dlock.go]: This key may conflict with meta or table record.

**Comment by ngaut** [comment]
I prefer to use etcd as distributed lock service. Which is more simple and clean.

**Comment by tiancaiamao** [comment]
I prefer etcd too, but it's not available currently. @ngaut 
Now we need some mechanism to guarantee bootstrap safety, with distributed lock, it should work.

Bootstrap check the code version and compare it with data version, 
if code version is greater, TiDB need upgrade the data.

Multiple TiDB instance may run bootstrap code concurrently. 
Safety is achieved by running update operation and update version a transaction, 
the process is atomic:

```
BEGIN
doSomeDML
updateVersion
END
```

But when upgrade involves DDL operation, the safety guarantee doesn't hold any more.

```
BEGIN
doSomeDML
doSomeDDL
updateVersion
END
```

or 

```
BEGIN
doSomeDDL
doSomeDML
updateVersion
END
```

because TiDB will autocommit DML as long as it meet DDL, that's not in a transaction any more.

File: kv/dlock/dlock.go (hunk 1 above)
Comment (tiancaiamao) [inline kv/dlock/dlock.go]: name is specified by the user, we can choose a name that doesn't conflict, such as "bootstrap".

**Comment by siddontang** [comment]
> I prefer etcd too, but it's not available currently

Why not available?

**Comment by tiancaiamao** [comment]
OK, I'll introduce etcd. @siddontang

**Pull request closed without merging**